Thread TLV encoding/decoding for mesh link establishment

mog96 opened the pull request:
This adds a single module, capsules/src/net/tlv.rs, containing the minimum TLV implementation necessary to support mesh link establishment (MLE) for a Sleepy End Device (SED) connecting to a Thread network.

Follows Chapter 4 of the Thread 1.1.1 Specification.

ppannuto commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+
+#[repr(u8)]
+pub enum PrefixSubTlvType {
+    HasRoute = 0,
```
Same here for `1`?

mog96 replied:
See previous comment.

ppannuto commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+
+pub struct HasRouteTlvValue {
+    r_border_router_16: u16,
+    r_preference: u8,
```
what's the `r_` here, `route`? We tend to prefer verbose names where they make sense

mog96 replied:
R_, P_, and S_ are used in the spec. It's unclear what exactly they stand for, but I think they're there to distinguish between the same data format being used for different purposes.

I put comments pointing to where in the spec the values with these prefixes are defined.

ppannuto commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+pub struct BorderRouterTlvValue {
+    // See 15.8.3
+    p_border_router_16: u16,
+    p_bits: u16,
```
Same question for `p_` here

mog96 replied:
See above

ppannuto commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+pub enum ServiceSubTlv {
+    Server {
+        s_server_16: u16,
+        s_server_data: [u8; MAX_VALUE_LENGTH],
```
And `s_`

mog96 replied:
See previous

ppannuto commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+}
+
+#[repr(u8)]
+pub enum ServiceSubTlvType {
```
any meaningful gaps here?

mog96 replied:
See first gap comment

ppannuto commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+    C = 0b0000010000000000,
+    R = 0b0000001000000000,
+    O = 0b0000000100000000,
+    N = 0b0000000010000000,
```
Could we give these meaningful names, or at a minimum add comments here expanding on what the letters stand for?

mog96 replied:
I figured anyone using this code to manage TLVs will be reading the spec as they do it, and these letters are defined there.

Added comments denoting what each stands for.

ppannuto commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+    N = 0b01000000,
+    R = 0b00100000,
+    C = 0b00010000,
+    B = 0b00001000,
```
Same comment here (meaningful names)

mog96 replied:
See first gap comment

ppannuto commented:
One other comment from @bradjc: You should rename the file to be a bit better scoped, something like `net/thread/tlv.rs` or `net/thread-tlv.rs` or `net/tlv-thread.rs`. I think I like the idea of a thread module in the net capsule the best (i.e. the first), but don't have strong feelings

mog96 commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+    }
+
+    // Returns PrefixSubTlv and true if stable, false otherwise.
+    pub fn decode(buf: &[u8]) -> SResult<Option<(PrefixSubTlv, bool)>> {
```
Fixed: Removed Option from all the return types. I was using it to return None in case the TLV type is not detected. Updated code handles it better, per a suggestion from Hubert.

The bool is to indicate whether the TLV data is 'stable' -- a status that's outlined in the spec.

mog96 commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+
+#[repr(u8)]
+pub enum NetworkDataTlvType {
+    Prefix = 1,
```
The annoying thing about these type numbers is that the ones missing here are filled by the Prefix and Service sub-TLVs. Added a comment explaining this.

mog96 commented on capsules/src/net/tlv.rs:
```diff
@@ -0,0 +1,1324 @@
+
+#[repr(u8)]
+pub enum LinkMode {
+    ReceiverOnWhenIdle = 0b00001000,
```
Nice nit. Fixed.

mog96 commented:
Updated to address your comments.

A side note: It appears the cargo format script doesn't like comments at the end of an enum, which I used in a few places to denote missing fields. I moved these comments to fix the issue. But might be something worth looking into.

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+const TL_WIDTH: usize = 2; // Type and length fields of TLV are each one byte.
+const MAX_VALUE_LENGTH: usize = 128; // Assume a TLV value will be no longer than 128 bytes.
+
+pub enum Tlv<'a> {
```
All of the public facing types and methods need documentation. Also, note that none of these variants are marked `pub`, which means a client module wouldn't be able to construct a `Tlv` directly. That might be on purpose, in which case something like the word "opaque" should appear in the comment explaining this.

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+}
+
+impl<'a> Tlv<'a> {
+    pub fn encode(&self, buf: &mut [u8]) -> SResult {
```
Perhaps a more descriptive name would be "serialize". How large of a buffer should a client pass in? what happens if it's not big enough? too big?

bbbert replied:
I'm not sure I agree with this. `encode` keeps the code uniform with respect to the ieee802154 code, which uses the same set of macros I used for header encoding/decoding. The behavior when the buffer is not big enough is also already specified by the `SResult` type. There is also no real set length for TLVs because the set of TLVs vary wildly depending on which MLE message is being sent. In other words, the responsibility for determining how much space to allocate falls on whoever is making the MLE message.

alevy replied:
OK. That's fine too

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+        stream_done!(TL_WIDTH)
+    }
+
+    pub fn decode(buf: &mut [u8]) -> SResult<Tlv> {
```
"deserialize" if "encode -> decode". Are there any expectations for passed in buffer (e.g. "the buffer should contain a TLV header formatted following the specifications in blah blah blah")? Why is it `&mut` rather than `&`?

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+}
+
+#[repr(u8)]
+pub enum TlvType {
```
Why is this type public if it's variants are not public?

alevy replied:
Also docs

mog96 replied:
I think variants are implicitly public. Tried to make them pub and got an error.

bbbert replied:
@alevy @mog96 There was some PR that removed the visibility from Rust enum variants: [https://github.com/rust-lang/rust/pull/28442](https://github.com/rust-lang/rust/pull/28442)

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+}
+
+#[repr(u8)]
+pub enum LinkMode {
```
Ditto.

For all of these enums, a brief comment on each of the varients summarizing their purpose would be useful. Probably just lifting and paraphrasing something from the spec.

mog96 replied:
I think the enum and field names are pretty implicit as-is. Especially if you're reading the spec along side the code.

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+}
+
+#[repr(u8)]
+pub enum MulticastResponder {
```
Doc

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+    EndDevice = 0b0100_0000,
+}
+
+// Used in Connectivity TLV
```
Double forward-slashes (`//`) don't show up in generated documentation, only triple slashes and `//!` (for module comments)

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+    Low = 0b1100_0000,
+}
+
+pub enum NetworkDataTlv<'a> {
```
Doc

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+}
+
+impl<'a> NetworkDataTlv<'a> {
+    pub fn encode(&self, buf: &mut [u8], stable: bool) -> SResult {
```
Same comments as for `Tlv#encode` + what does `stable` mean?

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+
+// Gaps in type numbers are fulled by PrefixSubTlv and ServiceSubTlv.
+#[repr(u8)]
+pub enum NetworkDataTlvType {
```
Doc

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+    }
+}
+
+pub enum PrefixSubTlv<'a> {
```
Doc

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+}
+
+impl<'a> PrefixSubTlv<'a> {
+    pub fn encode(&self, buf: &mut [u8], stable: bool) -> SResult {
```
Ditto

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1315 @@
+//! Implements Thread 1.1.1 Specification Type-Length-Value (TLV) encoding
+//! and decoding. Used during mesh link establishment (MLE), outlined in
+//! Chapter 4.
```
Rather than just point to the specification (which isn't even available without signing their pseudo-NDA), summarize what TLVs are, what MLE is used for and high-level how you might use this module to build (one part) of a thread implementation.

Doesn't have to be 6 pages or anything, just an executive summary. Someone actually implementing thread will likely have to read the spec anyway, but think of the audience being someone who's _reading_ our thread implementation and browsing this module to understand the details. The code is there, so they can read it for themselves, but what background to they need to understand what the code is doing?

mog96 commented:
Updated with documentation.

alevy commented on capsules/src/net/thread/tlv.rs:
```diff
@@ -0,0 +1,1411 @@
+
+impl<'a> Tlv<'a> {
+    /// Serializes TLV data in `buf` into the format specific to the TLV
+    /// type.
```
My understanding from reading the module-level comment is that these `&mut [u8]` buffers should be at least 128 bytes long for the various `encode` methods to always run successfully. Is that corrent?

mog96 replied:
I'm not sure we can answer that question just yet..

128 is an arbitrary number that I chose as the maximum length that a single field in a TLV value can have. It doesn't have to do with `encode` — it's important for `decode`, in cases where I have to decode a TLV value that can vary in length and stick it into a field in a `Tlv` variant to return to the caller of the decode function. For example, look at the `NetworkDataTlv::CommissioningData` variant. When I'm [decoding](https://github.com/ptcrews/tock/blob/31e8b12cf6eda86421dda39fc1bd0fee06591cab/capsules/src/net/thread/tlv.rs#L645) this type of TLV, according to the spec, the `com_data` field can vary in length, so ideally I'd want `com_data` to be a dynamically sized array. The way I would normally deal with that is by making `com_data` a reference to a slice of the buffer passed to `decode`. But because the contents of `com_data` must be decoded from network byte order before being returned to the user, I can't just return a slice of the original buffer. The solution I came up with was making the `com_data` field an array with a fixed length of 128, which I'm guessing will be sufficiently large to store the decoded contents.

That's where 128 comes into play in this module.

The buffer passed to `encode` must simply be large enough to store all of the fields of a given `Tlv` variant (which together represent the TLV value), plus two bytes for the TLV type and the TLV length. I was assuming that this buffer would simply be a slice of the buffer being used to encode the entire MLE of which this particular TLV is a part. It's unclear how large that MLE buffer should be; it's something we'll have to figure out when we implement MLE.
